client/app/services/widget.js

```diff
@@ -106,29 +106,27 @@ function WidgetFactory($http, Query, Visualization, dashboardGridOptions) {
     }
 
     load(force, maxAge) {
-      this.loading = true;
-      this.refreshStartedAt = moment();
-
       if (!this.visualization) {
         return undefined;
       }
 
-      if (force || this.queryResult === undefined) {
+      // Both `this.data` and `this.queryResult` are query result objects;
+      // `this.data` is last loaded query result;
+      // `this.queryResult` is currently loading query result;
+      // while widget is refreshing, `this.data` !== `this.queryResult`
+
+      if (force || (this.queryResult === undefined)) {
         if (maxAge === undefined || force) {
           maxAge = force ? 0 : undefined;
         }
         this.queryResult = this.getQuery().getQueryResult(maxAge);
+        this.loading = true;
+        this.refreshStartedAt = moment();
 
-        this.queryResult.toPromise().then(
-          (queryResult) => {
-            this.data = queryResult;
-            this.loading = false;
-          },
-          () => {
-            this.loading = false;
-            this.data = null;
-          },
-        );
+        this.queryResult.toPromise().finally(() => {
```

Comment:
Wouldn't it be cleaner to reject the Promise with a `QueryResultError` object, and then here use a local reference to query result instead of relaying on having a reference to the query result object that triggered the process?

I think it will make it easier to refactor into promises only API later on.

Reply:
I was thinking about it. But if we'll decide to reactor it - it may require changes in other places that use query results. For sure, I can do it, but it may take some time (including testing).

Reply:
But until now we were not rejecting this promise in case of errors, so no other code depends on it being of specific type. Or did you mean something else?

Reply:
Actually, all code related to query results does not rely on state of promise; instead it's using `getStatus()`/`getError()` method of query result, which can be `QueryResult` or `QueryResultError` instance. Maybe the right way is to reject promise when there are errors with query result, but there are some issues. First of all, when you request query result, all its methods will modify original object. Therefore it's hard to return new  `QueryResult`/`QueryResultError` instance in `then/catch`. Another issue is that `Query.getQueryResult()` will return query result instance, not a promise; that means, if `QueryResult` will receive status `4/failed` it should turn into `QueryResultError`, which is impossible (without changing a way we return query results for a query). That's what I mean when saying that it may require massive refactoring for `Query`, `QueryResults`, `Widget` services and related UI code.

Reply:
I wasn't suggesting a global refactor or change... I was referring to the local code in `services/widget.js`. Previously it was using the QueryResult object from the Promise:

https://github.com/getredash/redash/pull/2664/files#diff-3fe10675d2975aacb62dabd70951b589L124

But now, it uses the local `this.queryResult`: 

https://github.com/getredash/redash/pull/2664/files#diff-3fe10675d2975aacb62dabd70951b589R128

My suggestion is to resolve the promise with a `QueryResultError` object (instead of a string) and keep using the resolved object as the value we assign to `this.data` -- in the error state as well.